prow/config/config.go

```diff
@@ -499,16 +508,137 @@ type Plank struct {
+	// entries.
+	Config *prowapi.DecorationConfig `json:"config,omitempty"`
+}
+
+// Matches returns true iff all the filters for the entry match a job.
+func (d *DefaultDecorationConfigEntry) Matches(repo, cluster string) bool {
+	return d.Repo.MatchString(repo) && d.Cluster.MatchString(cluster)
+}
+
+// MergeDefaultDecorationConfig finds all matching DefaultDecorationConfigEntry
+// for a job and merges them sequentially before merging into the job's own
+// DecorationConfig. Configs merged later override values from earlier configs.
+func (p *Plank) MergeDefaultDecorationConfig(repo, cluster string, jobDC *prowapi.DecorationConfig) *prowapi.DecorationConfig {
+	var merged *prowapi.DecorationConfig
+	for _, entry := range p.DefaultDecorationConfigs {
+		if entry.Matches(repo, cluster) {
+			merged = entry.Config.ApplyDefault(merged)
+		}
+	}
+	merged = jobDC.ApplyDefault(merged)
+	if merged == nil {
+		merged = &prowapi.DecorationConfig{}
+	}
+	return merged
+}
+
+// defaultDecorationMapToSlice converts the old DefaultDecorationConfigs format:
+// map[string]*prowapi.DecorationConfig) to the new format: []*DefaultDecorationConfigEntry
+func defaultDecorationMapToSlice(m map[string]*prowapi.DecorationConfig) []*DefaultDecorationConfigEntry {
+	var entries []*DefaultDecorationConfigEntry
+	add := func(repo string, dc *prowapi.DecorationConfig) {
+		entries = append(entries, &DefaultDecorationConfigEntry{
+			RepoRaw:    repo,
+			ClusterRaw: ".*",
+			Config:     dc,
+		})
+	}
+	// Ensure "*" comes first...
+	if dc, exists := m["*"]; exists {
+		add(".*", dc)
+	}
+	// then orgs...
+	for key, dc := range m {
+		if key == "*" || strings.Contains(key, "/") {
+			continue
+		}
+		add(fmt.Sprintf("^%s/.*", key), dc)
 	}
-	org := strings.Split(repo, "/")[0]
-	if dcByOrg, ok := p.DefaultDecorationConfigs[org]; ok {
-		return dcByOrg.ApplyDefault(def)
+	// then repos.
+	for key, dc := range m {
+		if key == "*" || !strings.Contains(key, "/") {
+			continue
+		}
+		add(fmt.Sprintf("^%s$", key), dc)
 	}
-	return def
+	return entries
+}
+
+func compileDefaultDecorationRegex(entries []*DefaultDecorationConfigEntry) error {
+	var errs []error
+	for i, entry := range entries {
+		var err error
+		if entry.Repo, err = regexp.Compile(entry.RepoRaw); err != nil {
+			err = fmt.Errorf("compiling repo regex for plank.default_decoration_configs[%d]: %w", i, err)
+			errs = append(errs, err)
+		}
+		if entry.Cluster, err = regexp.Compile(entry.ClusterRaw); err != nil {
+			err = fmt.Errorf("compiling cluster regex for plank.default_decoration_configs[%d]: %w", i, err)
+			errs = append(errs, err)
+		}
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+// DefaultDecorationMapToSliceTesting is a convenience function that is exposed
+// to allow unit tests to convert the old map format to the new slice format.
+// It should only be used in testing.
+func DefaultDecorationMapToSliceTesting(m map[string]*prowapi.DecorationConfig) []*DefaultDecorationConfigEntry {
```

Comment:
Nit: Can we have a `Must` somewhere in the name so it becomes clear that this might panic?

Reply:
This can no longer panic after removing the regexp compilation.